lib/fog/vcloud_director/generators/compute/compose_common.rb

```diff
@@ -115,6 +122,68 @@ def build_source_items(xml)
             xml.AllEULAsAccepted (@configuration[:AllEULAsAccepted] || true)
           end
 
+          def build_virtual_hardware_section(xml, hardware)
+            xml[:ovf].VirtualHardwareSection do
+              xml[:ovf].Info('Virtual hardware requirements')
+              virtual_hardware_section_item_mem(xml, **hardware[:memory]) if hardware[:memory]
+              virtual_hardware_section_item_cpu(xml, **hardware[:cpu]) if hardware[:cpu]
+              if hardware[:disk]
+                hardware[:disk].each { |disk| virtual_hardware_section_item_hdd(xml, **disk) }
+              end
+            end
+          end
+
+          def virtual_hardware_section_item_base(xml, type:, name: nil, id: nil)
+            id ||= SecureRandom.uuid
+            name ||= id
+            xml[:ovf].Item do
+              xml[:rasd].ResourceType(type)
+              xml[:rasd].ElementName(name)
+              xml[:rasd].InstanceID(id)
```

Comment:
Some do not pass `id`. What happens when it’s nil?

Reply:
It gets assigned random UUID in L137. In most cases this `InstanceID` element makes no sense but vCD refuses the whole request if not passed. Except for disks: the `InstanceID` there specifies what disk are we resizing, hence we pass the `id` through.